python_modules/libraries/dagster-airbyte/dagster_airbyte_tests/experimental/test_resources.py

```diff
@@ -144,3 +158,226 @@ def test_basic_resource_request(
+    "status",
+    [
+        AirbyteJobStatusType.SUCCEEDED,
+        AirbyteJobStatusType.CANCELLED,
+        AirbyteJobStatusType.ERROR,
+        AirbyteJobStatusType.FAILED,
+        TEST_UNRECOGNIZED_AIRBYTE_JOB_STATUS_TYPE,
+    ],
+    ids=[
+        "job_status_succeeded",
+        "job_status_cancelled",
+        "job_status_error",
+        "job_status_failed",
+        "job_status_unrecognized",
+    ],
+)
+def test_airbyte_sync_and_poll_client_job_status(
+    status: str, base_api_mocks: responses.RequestsMock
+) -> None:
+    resource = AirbyteCloudWorkspace(
+        workspace_id=TEST_WORKSPACE_ID,
+        client_id=TEST_CLIENT_ID,
+        client_secret=TEST_CLIENT_SECRET,
+    )
+    client = resource.get_client()
+
+    test_job_endpoint = f"jobs/{TEST_JOB_ID}"
+    test_job_api_url = f"{AIRBYTE_REST_API_BASE}/{AIRBYTE_REST_API_VERSION}/{test_job_endpoint}"
+
+    # Create mock responses to mock full sync and poll behavior to test statuses, used only in this test
+    base_api_mocks.add(
+        method=responses.POST,
+        url=f"{AIRBYTE_REST_API_BASE}/{AIRBYTE_REST_API_VERSION}/jobs",
+        json=get_job_details_sample(status=AirbyteJobStatusType.PENDING),
+        status=200,
+    )
+    base_api_mocks.add(
+        method=responses.GET,
+        url=test_job_api_url,
+        json=get_job_details_sample(status=status),
+        status=200,
+    )
+    base_api_mocks.add(
+        method=responses.POST,
+        url=f"{AIRBYTE_CONFIGURATION_API_BASE}/{AIRBYTE_CONFIGURATION_API_VERSION}/connections/get",
+        json=SAMPLE_CONNECTION_DETAILS,
+        status=200,
+    )
+
+    if status == TEST_UNRECOGNIZED_AIRBYTE_JOB_STATUS_TYPE:
+        base_api_mocks.add(
+            method=responses.DELETE,
+            url=test_job_api_url,
+            status=200,
+            json=get_job_details_sample(status=AirbyteJobStatusType.CANCELLED),
+        )
+
+    if status in [AirbyteJobStatusType.ERROR, AirbyteJobStatusType.FAILED]:
+        with pytest.raises(Failure, match="Job failed"):
+            client.sync_and_poll(connection_id=TEST_CONNECTION_ID, poll_interval=0)
+
+    elif status == AirbyteJobStatusType.CANCELLED:
+        with pytest.raises(Failure, match="Job was cancelled"):
+            client.sync_and_poll(connection_id=TEST_CONNECTION_ID, poll_interval=0)
+
+    elif status == TEST_UNRECOGNIZED_AIRBYTE_JOB_STATUS_TYPE:
+        with pytest.raises(Failure, match="unexpected state"):
+            client.sync_and_poll(connection_id=TEST_CONNECTION_ID, poll_interval=0)
+        assert_rest_api_call(
+            call=base_api_mocks.calls[-1], endpoint=test_job_endpoint, method=responses.DELETE
+        )
+
+    else:
+        result = client.sync_and_poll(connection_id=TEST_CONNECTION_ID, poll_interval=0)
+        assert result == AirbyteOutput(
+            job_details=get_job_details_sample(AirbyteJobStatusType.SUCCEEDED),
+            connection_details=SAMPLE_CONNECTION_DETAILS,
+        )
```

Comment:
nit; this could all be part of the parameterization. I actually had to do something similar recently, and I found an `optional_pytest_raise` to be really useful here: https://github.com/dagster-io/dagster/pull/26545/files#diff-bb4b77564796719ee9a5eaf83ecd67f54f4f16b8b7724f9348d1a38dac76800d

Reply:
Updated in [5df711b](https://github.com/dagster-io/dagster/pull/26431/commits/5df711bd399de0494fed1d6a6ad2e0296072ac1c)